// @markuplint/ml-core/lib/ml-dom/tokens/omitted-element.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const node_1 = tslib_1.__importDefault(require("./node"));
class MLDOMOmittedElement extends node_1.default {
    constructor(astNode, document) {
        super(astNode, document);
        this.type = 'OmittedElement';
        this.obsolete = false;
        this.nodeName = astNode.nodeName;
        this.namespaceURI = astNode.namespace;
        this.isForeignElement = this.namespaceURI !== 'http://www.w3.org/1999/xhtml';
    }
    get childNodes() {
        const astChildren = this._astToken.childNodes || [];
        return astChildren.map(node => this.nodeStore.getNode(node));
    }
    getChildElementsAndTextNodeWithoutWhitespaces() {
        const filteredNodes = [];
        this.childNodes.forEach(node => {
            if (node.type === 'Element') {
                filteredNodes.push(node);
            }
            if (node.type === 'Text' && !node.isWhitespace()) {
                filteredNodes.push(node);
            }
            if (node.type === 'OmittedElement') {
                const children = node.getChildElementsAndTextNodeWithoutWhitespaces();
                filteredNodes.push(...children);
            }
        });
        return filteredNodes;
    }
}
exports.default = MLDOMOmittedElement;
