// @markuplint/rules/lib/required-h1/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
exports.default = ml_core_1.createRule({
    name: 'required-h1',
    defaultValue: true,
    defaultOptions: {
        'expected-once': true,
        'in-document-fragment': false,
    },
    async verify(document, translate, globalRule) {
        const reports = [];
        const h1Stack = [];
        if (document.nodeList.length === 0) {
            return reports;
        }
        if (!globalRule.option['in-document-fragment'] && document.isFragment) {
            return reports;
        }
        await document.walkOn('Element', async (node) => {
            if (node.nodeName.toLowerCase() === 'h1') {
                h1Stack.push(node);
            }
        });
        if (h1Stack.length === 0) {
            const message = translate('Missing the {0} {1}', 'h1', 'element');
            reports.push({
                severity: globalRule.severity,
                message,
                line: 1,
                col: 1,
                raw: document.nodeList[0].raw.slice(0, 1),
            });
        }
        else if (globalRule.option['expected-once'] && h1Stack.length > 1) {
            const message = translate('Duplicate the {0} {1}', 'h1', 'element');
            reports.push({
                severity: globalRule.severity,
                message,
                line: h1Stack[1].startLine,
                col: h1Stack[1].startCol,
                raw: h1Stack[1].raw,
            });
        }
        return reports;
    },
});
