// @markuplint/rules/lib/id-duplication/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
exports.default = ml_core_1.createRule({
    name: 'id-duplication',
    defaultValue: null,
    defaultOptions: null,
    async verify(document, translate) {
        const reports = [];
        const message = translate('Duplicate {0}', 'attribute id value');
        const idStack = [];
        await document.walkOn('Element', async (node) => {
            const idAttrs = node.getAttributeToken('id');
            for (const idAttr of idAttrs) {
                if (idAttr.attrType === 'ps-attr' ||
                    (idAttr.attrType === 'html-attr' && idAttr.isDynamicValue) ||
                    (idAttr.attrType === 'html-attr' && idAttr.isDirective)) {
                    continue;
                }
                const id = idAttr.getValue();
                if (idStack.includes(id.raw)) {
                    reports.push({
                        severity: node.rule.severity,
                        message,
                        line: idAttr.startLine,
                        col: idAttr.startCol,
                        raw: idAttr.raw,
                    });
                }
                idStack.push(id.raw);
            }
        });
        return reports;
    },
});
