// @markuplint/rules/lib/deprecated-attr/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
const helpers_1 = require("../helpers");
exports.default = ml_core_1.createRule({
    name: 'deprecated-attr',
    defaultValue: null,
    defaultOptions: null,
    async verify(document, translate) {
        const reports = [];
        await document.walkOn('Element', async (element) => {
            const attrSpecs = helpers_1.getAttrSpecs(element.nodeName, document.specs);
            if (!attrSpecs) {
                return;
            }
            for (const attr of element.attributes) {
                const name = attr.getName();
                const attrSpec = attrSpecs.find(item => item.name === name.potential);
                if (!attrSpec) {
                    return;
                }
                if (attrSpec.deprecated || attrSpec.obsolete) {
                    const message = translate('The {0} {1} is {2}', name.potential, 'attribute', attrSpec.obsolete ? 'obsolete' : 'deprecated');
                    reports.push({
                        severity: element.rule.severity,
                        message,
                        line: name.line,
                        col: name.col,
                        raw: name.raw,
                    });
                }
            }
        });
        return reports;
    },
});
