// @markuplint/rules/lib/deprecated-element/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const tslib_1 = require("tslib");
const ml_core_1 = require("@markuplint/ml-core");
const ml_spec_1 = require("@markuplint/ml-spec");
const html_spec_1 = tslib_1.__importDefault(require("@markuplint/html-spec"));
exports.default = ml_core_1.createRule({
    name: 'deprecated-element',
    defaultValue: null,
    defaultOptions: null,
    async verify(document, translate) {
        const reports = [];
        const message = translate('{0} is {1}', 'Element', 'deprecated');
        await document.walkOn('Element', async (element) => {
            if (element.isForeignElement || element.isCustomElement) {
                return;
            }
            const spec = ml_spec_1.getSpecByTagName(element.nodeName, html_spec_1.default);
            if (spec && (spec.obsolete || spec.deprecated || spec.nonStandard)) {
                reports.push({
                    severity: element.rule.severity,
                    message,
                    line: element.startLine,
                    col: element.startCol,
                    raw: element.raw,
                });
            }
        });
        return reports;
    },
});
