// markuplint/lib/cli-bootstrap.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.cli = exports.help = void 0;
const tslib_1 = require("tslib");
const meow_1 = tslib_1.__importDefault(require("meow"));
exports.help = `
Usage
	$ markuplint <HTML file pathes (glob format)>
	$ <stdout> | markuplint

Options
	--config-file,  -c FILE_PATH  Ruleset file path.
	--fix,                        Fix HTML.
	--format,       -f FORMAT     Output format. Support "JSON", "Simple" and "Standard". Default: "Standard".
	--no-color,                   Output no color.
	--problem-only, -p            Output only problems, without passeds.
	--verbose                     Output with detailed information.

	--init                        Initialize settings interactively.

	--help,         -h            Show help.
	--version,      -v            Show version.

Examples
	$ markuplint verifyee.html --ruleset path/to/.markuplintrc
	$ cat verifyee.html | markuplint
`;
exports.cli = meow_1.default(exports.help, {
    flags: {
        init: {
            type: 'boolean',
        },
        configFile: {
            type: 'string',
            alias: 'c',
        },
        fix: {
            type: 'boolean',
        },
        format: {
            type: 'string',
            alias: 'f',
        },
        color: {
            type: 'boolean',
            default: true,
        },
        problemOnly: {
            type: 'boolean',
            alias: 'p',
        },
        verbose: {
            type: 'boolean',
        },
    },
});
