// @markuplint/rules/lib/attr-duplication/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const ml_core_1 = require("@markuplint/ml-core");
exports.default = ml_core_1.createRule({
    name: 'attr-duplication',
    defaultValue: null,
    defaultOptions: null,
    async verify(document, translate) {
        const reports = [];
        const message = translate('Duplicate {0}', 'attribute name');
        await document.walkOn('Element', async (node) => {
            const attrNameStack = [];
            for (const attr of node.attributes) {
                if (attr.isDuplicatable) {
                    continue;
                }
                const attrName = attr.getName();
                const name = node.isCustomElement ? attrName.potential : attrName.potential.toLowerCase();
                if (attrNameStack.includes(name)) {
                    reports.push({
                        severity: node.rule.severity,
                        message,
                        line: attrName.line,
                        col: attrName.col,
                        raw: attrName.raw,
                    });
                }
                else {
                    attrNameStack.push(name);
                }
            }
        });
        return reports;
    },
});
